runtime/win32.c

```diff
@@ -96,40 +96,40 @@ static void caml_win32_sys_error(int errnum)
 int caml_read_fd(int fd, int flags, void * buf, int n)
 {
   int retcode;
+  caml_enter_blocking_section_no_pending();
   if ((flags & CHANNEL_FLAG_FROM_SOCKET) == 0) {
-    caml_enter_blocking_section();
     retcode = read(fd, buf, n);
     /* Large reads from console can fail with ENOMEM.  Reduce requested size
        and try again. */
     if (retcode == -1 && errno == ENOMEM && n > 16384) {
       retcode = read(fd, buf, 16384);
     }
-    caml_leave_blocking_section();
-    if (retcode == -1) caml_sys_io_error(NO_ARG);
   } else {
-    caml_enter_blocking_section();
     retcode = recv((SOCKET) _get_osfhandle(fd), buf, n, 0);
-    caml_leave_blocking_section();
-    if (retcode == -1) caml_win32_sys_error(WSAGetLastError());
+    if (retcode == -1) {
+      errno = caml_posixerr_of_win32err(WSAGetLastError());
+      if (errno == 0) errno = EINVAL;
+    }
   }
+  caml_leave_blocking_section();
   return retcode;
 }
 
 int caml_write_fd(int fd, int flags, void * buf, int n)
 {
   int retcode;
+  caml_enter_blocking_section_no_pending();
   if ((flags & CHANNEL_FLAG_FROM_SOCKET) == 0) {
-    caml_enter_blocking_section();
     retcode = write(fd, buf, n);
-    caml_leave_blocking_section();
-    if (retcode == -1) caml_sys_io_error(NO_ARG);
   } else {
-    caml_enter_blocking_section();
     retcode = send((SOCKET) _get_osfhandle(fd), buf, n, 0);
-    caml_leave_blocking_section();
-    if (retcode == -1) caml_win32_sys_error(WSAGetLastError());
+    if (retcode == -1) {
+      errno = caml_posixerr_of_win32err(WSAGetLastError());
```

Comment:
`WSAGetLastError` has a [return type](https://learn.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-wsagetlasterror) of `int` (_vs_ the `unsigned int` expected by `caml_posixerr_of_win32err`). But reading through the [windows socket error codes](https://learn.microsoft.com/en-us/windows/win32/winsock/windows-sockets-error-codes-2) there doesn't seem to be any use of negative error codes.

Reply:
Thanks for having checked.  The `unsigned int` is to match the `DWORD` type returned by `GetLastError`.  I used `unsigned int` instead of `DWORD` so that `<windows.h>` doesn't have to be included from `caml/osdeps.h`.

Reply:
That makes sense, but you also call `WSAGetLastError()` in https://github.com/ocaml/ocaml/pull/12314/files#diff-a171e424a2878bac63e4d485d10e8137d1a81eb35aa7562047ce6a446ec17294R110, which is a compatibility wrapper these days over `GetLastError`. And for some reason, that compatibility wrapper has `int` as the return value.  You could just replace the `WSAGetLastError` with `GetLastError`, I think?

Reply:
I'm not sure what problem you're pointing out here.
- No code is generated when converting between `signed  int` and `unsigned int`, so the bit patterns of error codes are preserved when calling `caml_posixerr_of_win32err`, and that's what matters.  (Plus: Win32 and Winsock2 error codes are small positive integers anyway.)
- That `WSAGetLastError` returns `int` and not `DWORD` is strange, indeed, but we can live with it.
- I didn't know that `WSAGetLastError` is a wrapper around `GetLastError`.  At any rate, this is not documented in MSDN, so let's stick to the documented Winsock2 API and religiously call `WSAGetLastError` after a Winsock2 operation.

Reply:
>  so let's stick to the documented Winsock2 API and religiously call WSAGetLastError after a Winsock2 operation.

Right, but you're also calling `GetLastError()` in this diff (line 819).  I agree WSAGetLastError() is a better choice than GetLastError() due to being the official API.
Also agree that there's no safety problem here due to the Windows error codes being small and non-negative, but it makes the code harder to read to have both functions being used.

Reply:
The `GetLastError` on line 819 was there before and is appropriate because it follows Win32 system calls.  The rule we've always applied is that if it's a Win32 function, use `GetLastError`, and if it's a Winsock function, use `WSAGetLastError`.

Reply:
Ahhh, that makes sense...many thanks for the explanation. I'd entirely missed the distinction between Winsock and Win32 system calls.